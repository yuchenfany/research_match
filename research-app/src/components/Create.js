/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-filename-extension */
import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import '../assets/index.css';

function Create({ user, setUser }) {
  const [error, setError] = useState({ message: '' });
  const navigate = useNavigate();

  // combined alphanumeric and empty check
  const isValid = () => {
    if (user.name.length === 0) {
      setError({ message: 'Please enter a username' });
    } else if (!user.name.match(/^[0-9a-zA-Z]+$/)) {
      setError({ message: 'Username must be an alphanumeric string' });
    }
    return !(user.name.length === 0 || !user.name.match(/^[0-9a-zA-Z]+$/));
  };

  async function verify() {
    console.log(JSON.stringify(user));

    await fetch('http://localhost:5000/record/add', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(user),
    })
      .catch((e) => {
        window.alert(e);
      });
  }

  async function handleSubmit(event) {
    if (await verify()) {
      console.log('verified');
      navigate('/home');
    } else {
      event.preventDefault();
    }
  }

  // update username as it's being entered
  const handleNameChange = async (event) => {
    setUser(
      {
        username: event.target.value,
        password: user.password,
      },
    );
  };

  const handleNameChangePassword = async (event) => {
    setUser(
      {
        username: user.username,
        password: event.target.value,
      },
    );
  };

  const backToLogin = () => {
    navigate('/');
  };

  const handleAsync = (event) => {
    event.preventDefault();
    handleNameChange(event).then(handleSubmit(event));
  };

  return (
    <div className="Create">
      <p className="header">Sign Up</p>
      <form onSubmit={handleAsync}>
        <label className="login-label" htmlFor="username">
          <div className="username-wrapper">
            <p className="field-label">CREATE USERNAME</p>
            <input
              className="input-field"
              type="text"
              id="username"
              onChange={handleNameChange}
            />
            <p className="field-label">CREATE PASSWORD</p>
            <input
              className="input-field"
              type="text"
              id="password"
              onChange={handleNameChangePassword}
            />
            <span className="error-message">{error.message}</span>
          </div>
          <input className="button" type="submit" value="CREATE ACCOUNT" />
        </label>
        <div className="spacer" />
        <button className="link" type="button" onClick={backToLogin}>Back to login</button>
      </form>
    </div>
  );
}

export default Create;
