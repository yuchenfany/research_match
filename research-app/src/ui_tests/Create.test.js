/**
 * @jest-environment jsdom
 */

/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-filename-extension */
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import React from 'react';
// import { jest } from '@jest/globals';

import Create from '../components/Create';

// const mockedUsedNavigate = jest.fn();

// jest.mock('react-router-dom', () => ({
//   ...jest.requireActual('react-router-dom'),
//   useNavigate: () => mockedUsedNavigate,
// }));

test('Sign Up header', () => {
  render(
    <Router>
      <Create />
    </Router>,
  );
  const header = screen.getByText('Sign Up');
  expect(header).toBeInTheDocument();
});

test('CREATE USERNAME label', () => {
  render(
    <Router>
      <Create />
    </Router>,
  );
  const label = screen.getByText('CREATE USERNAME');
  expect(label).toBeInTheDocument();
});

test('CREATE PASSWORD label', () => {
  render(
    <Router>
      <Create />
    </Router>,
  );
  const label = screen.getByText('CREATE PASSWORD');
  expect(label).toBeInTheDocument();
});

test('CREATE ACCOUNT button is clicked', () => {
  const mockSetUser = jest.fn();
  const user1 = { username: 'user1', password: 'user1pass' };
  render(
    <Router>
      <Create setUser={mockSetUser} user={user1} />
    </Router>,
  );
  const input = screen.getByPlaceholderText('Create username');
  fireEvent.change(input, { target: { value: 'user1' } });
  const button = screen.getByRole('button', {
    name: /create account/i,
  });

  fireEvent.click(button);
  expect(mockSetUser).toBeCalled();
});

test('Back to login button is clicked: valid input', () => {
  const mockSetUser = jest.fn();
  const user1 = { username: 'user1', password: 'user1pass' };
  render(
    <Router>
      <Create setUser={mockSetUser} user={user1} />
    </Router>,
  );
  const input = screen.getByPlaceholderText('Create username');
  fireEvent.change(input, { target: { value: 'user1' } });

  const inputPassword = screen.getByPlaceholderText('Create password');
  fireEvent.change(inputPassword, { target: { value: 'user1pass' } });

  const button = screen.getByRole('button', {
    name: /back to login/i,
  });

  fireEvent.click(button);
  expect(mockSetUser).toBeCalled();
});

test('snapshot test', () => {
  const component = renderer.create(
    <Router>
      <Create />
    </Router>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
