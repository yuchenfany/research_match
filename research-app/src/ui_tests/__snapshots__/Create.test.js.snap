// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<div
  className="Create"
>
  <p
    className="header"
  >
    Sign Up
  </p>
  <form
    onSubmit={[Function]}
  >
    <label
      className="login-label"
      htmlFor="username"
    >
      <div
        className="username-wrapper"
      >
        <p
          className="field-label"
        >
          CREATE USERNAME
        </p>
        <input
          className="input-field"
          id="username"
          onChange={[Function]}
          placeholder="Create username"
          type="text"
        />
        <p
          className="field-label"
        >
          CREATE PASSWORD
        </p>
        <input
          className="input-field"
          id="password"
          onChange={[Function]}
          placeholder="Create password"
          type="text"
        />
        <span
          className="error-message"
        >
          
        </span>
      </div>
      <input
        className="button"
        type="submit"
        value="CREATE ACCOUNT"
      />
    </label>
    <div
      className="spacer"
    />
    <button
      className="link"
      onClick={[Function]}
      type="button"
    >
      Back to login
    </button>
  </form>
</div>
`;
